### PR TITLE
build: format c code with clang-format

### PR DESCRIPTION
**Pull request by dundargoc**
Add the make targets `lint` and `format` so developers can quickly test
their code is compliant locally. Also add check in CI to ensure repo is
always correctly formatted.

### CODE REVIEW

File: highlight/include/tree_sitter/highlight.h [hunk 1]
```diff
@@ -39,29 +36,17 @@ void ts_highlighter_delete(TSHighlighter *);
...
-  const TSLanguage *language,
-  const char *highlight_query,
-  const char *injection_query,
-  const char *locals_query,
-  uint32_t highlight_query_len,
-  uint32_t injection_query_len,
-  uint32_t locals_query_len,
-  bool apply_all_captures
-);
+  TSHighlighter *self, const char *scope_name, const char *injection_regex,
```
Comment (lewis6991) [inline highlight/include/tree_sitter/highlight.h:39]: This is a big regression IMO. Can we fix this?

**Comment by dundargoc** [comment]
clang-format can't have one function parameter on each line. The technology isn't there yet. I'm hesitant to introduce uncrustify, but maybe that's still better than no formatter. Unsure. Closing for now.

**Comment by lewis6991** [comment]
Gemini thinks this is possible now:

---

Yes, clang-format can be configured to format functions with many parameters so that each parameter is on a single line. This is a common requirement for readability, especially when parameters have long names or complex types.
Here are the key clang-format options you'll want to configure in your .clang-format file to achieve this:
 * BinPackArguments: false: This option prevents arguments (in function calls) from being "bin-packed" or placed on the same line if they fit within the ColumnLimit. Setting it to false encourages clang-format to break arguments onto new lines more readily.
 * BinPackParameters: false: Similar to BinPackArguments, this option applies to function declarations/definitions. Setting it to false will prevent parameters from being kept on the same line if they would otherwise fit, pushing them to new lines.
 * AllowAllParametersOfDeclarationOnNextLine: false: This option, when false, helps ensure that if a function declaration or definition breaks, all parameters go to new lines (rather than having the first parameter stay on the same line as the function name).
 * BreakBeforeParameterName: false (or related options like BreakBeforeParameterName: Never): While less common for forcing all parameters to new lines, this option can influence how a single parameter is broken if it itself is very long. Generally, for the "one parameter per line" style, you'll be more focused on the BinPack options.
 * AlignAfterOpenBracket: BlockIndent (or DontAlign): This option controls how arguments are aligned after the opening parenthesis. BlockIndent will indent subsequent arguments to the same level as the first argument, while DontAlign will just follow the normal indentation. For one parameter per line, BlockIndent often looks cleaner.
 * ColumnLimit: While not directly forcing each parameter to a new line, ColumnLimit plays a significant role. If your ColumnLimit is very high, clang-format might still try to put multiple parameters on a single line even with BinPackArguments: false and BinPackParameters: false. A reasonable ColumnLimit (e.g., 80 or 100) helps trigger the line breaks for parameters.
Example .clang-format snippet for this style:
BasedOnStyle: LLVM # Or whatever base style you prefer
ColumnLimit: 80 # Adjust as needed
BinPackArguments: false
BinPackParameters: false
AllowAllParametersOfDeclarationOnNextLine: false
AlignAfterOpenBracket: BlockIndent

Important Notes:
 * clang-format versions: Some options and their exact behavior can vary slightly between clang-format versions. Always refer to the documentation for your specific clang-format version.
 * Trial and error: Formatting can be subjective. You might need to experiment with these and other clang-format options to achieve the precise look you desire.
 * Base Style: Starting with a BasedOnStyle (like LLVM, Google, Chromium, etc.) and then overriding specific options is generally a good approach.
 * .clang-format file: Place a .clang-format file with your desired configuration in the root of your project. clang-format will automatically discover and apply it.
By carefully configuring these options, you can effectively force clang-format to put each function parameter on its own line when you have a function with many parameters.

---

Hallucination or not?

**Comment by dundargoc** [comment]
Continued in https://github.com/tree-sitter/tree-sitter/pull/4517

**Pull request closed without merging**